File: src/hab/client/rest.py
"""REST API client for Home Assistant."""

from __future__ import annotations

from typing import Any

import httpx

from hab.exceptions import (
    AuthenticationError,
    ConnectionError,
    ResourceNotFoundError,
    TimeoutError,
    ValidationError,
)


class RestClient:
    """HTTP client for Home Assistant REST API."""

    def __init__(
        self,
        url: str,
        token: str,
        timeout: int = 30,
        verify_ssl: bool = True,
    ) -> None:
        """Initialize the REST client.

        Args:
            url: Home Assistant URL.
            token: Access token.
            timeout: Request timeout in seconds.
            verify_ssl: Whether to verify SSL certificates.
        """
        self.base_url = url.rstrip("/")
        self.token = token
        self.timeout = timeout
        self.verify_ssl = verify_ssl
        self._client: httpx.AsyncClient | None = None

    @property
    def client(self) -> httpx.AsyncClient:
        """Get or create the HTTP client."""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(
                base_url=self.base_url,
                headers={
                    "Authorization": f"Bearer {self.token}",
                    "Content-Type": "application/json",
                },
                timeout=self.timeout,
                verify=self.verify_ssl,
            )
        return self._client

    async def close(self) -> None:
        """Close the HTTP client."""
        if self._client:
            await self._client.aclose()
            self._client = None

    async def __aenter__(self) -> "RestClient":
        """Async context manager entry."""
        return self

    async def __aexit__(self, *args: Any) -> None:
        """Async context manager exit."""
        await self.close()

    def _handle_error(self, response: httpx.Response) -> None:
        """Handle HTTP error responses.

        Args:
            response: HTTP response to check.

        Raises:
            Appropriate exception based on status code.
        """
        if response.is_success:
            return

        try:
            data = response.json()
            message = data.get("message", response.text)
        except Exception:
            message = response.text

        if response.status_code == 401:
            raise AuthenticationError(f"Authentication failed: {message}")
        elif response.status_code == 403:
            from hab.exceptions import PermissionError
            raise PermissionError(f"Permission denied: {message}")
        elif response.status_code == 404:
            raise ResourceNotFoundError(f"Not found: {message}")
        elif response.status_code == 400:
            raise ValidationError(f"Bad request: {message}")
        else:
            raise ConnectionError(f"API error ({response.status_code}): {message}")

    async def _request(
        self,
        method: str,
        endpoint: str,
        **kwargs: Any,
    ) -> Any:
        """Make an HTTP request.

        Args:
            method: HTTP method.
            endpoint: API endpoint.
            **kwargs: Additional request arguments.

        Returns:
            Response data.

        Raises:
            Various exceptions based on response.
        """
        url = f"/api/{endpoint.lstrip('/')}"

        try:
            response = await self.client.request(method, url, **kwargs)
            self._handle_error(response)

            if response.status_code == 204:
                return None

            if response.headers.get("content-type", "").startswith("application/json"):
                return response.json()

            return response.text

        except httpx.TimeoutException as e:
            raise TimeoutError(f"Request timed out: {e}")
        except httpx.ConnectError as e:
            raise ConnectionError(f"Failed to connect: {e}")

    async def get(self, endpoint: str, **kwargs: Any) -> Any:
        """Make a GET request.

        Args:
            endpoint: API endpoint.
            **kwargs: Additional request arguments.

        Returns:
            Response data.
        """
        return await self._request("GET", endpoint, **kwargs)

    async def post(self, endpoint: str, data: Any = None, **kwargs: Any) -> Any:
        """Make a POST request.

        Args:
            endpoint: API endpoint.
            data: JSON body data.
            **kwargs: Additional request arguments.

        Returns:
            Response data.
        """
        return await self._request("POST", endpoint, json=data, **kwargs)

    async def put(self, endpoint: str, data: Any = None, **kwargs: Any) -> Any:
        """Make a PUT request.

        Args:
            endpoint: API endpoint.
            data: JSON body data.
            **kwargs: Additional request arguments.

        Returns:
            Response data.
        """
        return await self._request("PUT", endpoint, json=data, **kwargs)

    async def delete(self, endpoint: str, **kwargs: Any) -> Any:
        """Make a DELETE request.

        Args:
            endpoint: API endpoint.
            **kwargs: Additional request arguments.

        Returns:
            Response data.
        """
        return await self._request("DELETE", endpoint, **kwargs)

    # High-level API methods

    async def get_config(self) -> dict[str, Any]:
        """Get Home Assistant configuration."""
        return await self.get("config")

    async def get_states(self) -> list[dict[str, Any]]:
        """Get all entity states."""
        return await self.get("states")

    async def get_state(self, entity_id: str) -> dict[str, Any]:
        """Get state for a specific entity.

        Args:
            entity_id: Entity ID.

        Returns:
            Entity state.
        """
        return await self.get(f"states/{entity_id}")

    async def set_state(
        self,
        entity_id: str,
        state: str,
        attributes: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Set state for an entity.

        Args:
            entity_id: Entity ID.
            state: New state value.
            attributes: Optional attributes.

        Returns:
            Updated entity state.
        """
        data = {"state": state}
        if attributes:
            data["attributes"] = attributes
        return await self.post(f"states/{entity_id}", data)

    async def get_services(self) -> list[dict[str, Any]]:
        """Get all available services."""
        return await self.get("services")

    async def call_service(
        self,
        domain: str,
        service: str,
        data: dict[str, Any] | None = None,
        return_response: bool = False,
    ) -> Any:
        """Call a service.

        Args:
            domain: Service domain.
            service: Service name.
            data: Service data.
            return_response: Whether to return the service response.

        Returns:
            Service response if return_response is True.
        """
        if data is None:
            data = {}
        if return_response:
            data["return_response"] = True
        return await self.post(f"services/{domain}/{service}", data)

    async def get_history(
        self,
        entity_id: str | None = None,
        start_time: str | None = None,
        end_time: str | None = None,
    ) -> list[list[dict[str, Any]]]:
        """Get state history.

        Args:
            entity_id: Optional entity ID to filter.
            start_time: Start time (ISO format).
            end_time: End time (ISO format).

        Returns:
            History data.
        """
        params: dict[str, str] = {}
        if entity_id:
            params["filter_entity_id"] = entity_id
        if end_time:
            params["end_time"] = end_time

        endpoint = "history/period"
        if start_time:
            endpoint = f"history/period/{start_time}"

        return await self.get(endpoint, params=params)

    async def render_template(self, template: str) -> str:
        """Render a Jinja2 template.

        Args:
            template: Template string.

        Returns:
            Rendered template.
        """
        return await self.post("template", {"template": template})

    async def check_config(self) -> dict[str, Any]:
        """Check Home Assistant configuration.

        Returns:
            Configuration check result.
        """
        return await self.post("config/core/check_config")

    async def restart(self) -> None:
        """Restart Home Assistant."""
        await self.post("services/homeassistant/restart")

    async def get_error_log(self) -> str:
        """Get error log.

        Returns:
            Error log content.
        """
        return await self.get("error_log")

File: client/rest.go
package client

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/go-resty/resty/v2"
	log "github.com/sirupsen/logrus"
)

const (
	// DefaultTimeout is the default request timeout
	DefaultTimeout = 30 * time.Second
)

// RestClient is an HTTP client for the Home Assistant REST API
type RestClient struct {
	BaseURL   string
	Token     string
	Timeout   time.Duration
	VerifySSL bool
	client    *resty.Client
}

// NewRestClient creates a new REST client
func NewRestClient(baseURL, token string) *RestClient {
	return &RestClient{
		BaseURL:   baseURL,
		Token:     token,
		Timeout:   DefaultTimeout,
		VerifySSL: true,
	}
}

// NewRestClientWithOptions creates a REST client with custom options
func NewRestClientWithOptions(baseURL, token string, timeout time.Duration, verifySSL bool) *RestClient {
	return &RestClient{
		BaseURL:   baseURL,
		Token:     token,
		Timeout:   timeout,
		VerifySSL: verifySSL,
	}
}

func (c *RestClient) getClient() *resty.Client {
	if c.client == nil {
		c.client = resty.New()
		c.client.SetTimeout(c.Timeout)
		c.client.SetBaseURL(c.BaseURL)
		c.client.SetHeader("Authorization", "Bearer "+c.Token)
		c.client.SetHeader("Content-Type", "application/json")
		c.client.SetHeader("Accept", "application/json")

		if !c.VerifySSL {
			c.client.SetTLSClientConfig(&tls.Config{InsecureSkipVerify: true})
		}

		// Response logging
		c.client.OnAfterResponse(func(client *resty.Client, resp *resty.Response) error {
			log.WithFields(log.Fields{
				"status":     resp.StatusCode(),
				"time":       resp.Time(),
				"url":        resp.Request.URL,
				"method":     resp.Request.Method,
				"bodyLength": len(resp.Body()),
			}).Debug("REST response")
			return nil
		})
	}
	return c.client
}

// Get makes a GET request
func (c *RestClient) Get(endpoint string) (interface{}, error) {
	return c.request("GET", endpoint, nil)
}

// Post makes a POST request
func (c *RestClient) Post(endpoint string, body interface{}) (interface{}, error) {
	return c.request("POST", endpoint, body)
}

// Put makes a PUT request
func (c *RestClient) Put(endpoint string, body interface{}) (interface{}, error) {
	return c.request("PUT", endpoint, body)
}

// Delete makes a DELETE request
func (c *RestClient) Delete(endpoint string) (interface{}, error) {
	return c.request("DELETE", endpoint, nil)
}

func (c *RestClient) request(method, endpoint string, body interface{}) (interface{}, error) {
	url := fmt.Sprintf("/api/%s", endpoint)

	req := c.getClient().R()

	if body != nil {
		req.SetBody(body)
	}

	log.WithFields(log.Fields{
		"method": method,
		"url":    c.BaseURL + url,
	}).Debug("REST request")

	var resp *resty.Response
	var err error

	switch method {
	case "GET":
		resp, err = req.Get(url)
	case "POST":
		resp, err = req.Post(url)
	case "PUT":
		resp, err = req.Put(url)
	case "DELETE":
		resp, err = req.Delete(url)
	default:
		return nil, fmt.Errorf("unsupported method: %s", method)
	}

	if err != nil {
		return nil, fmt.Errorf("request failed: %w", err)
	}

	return c.handleResponse(resp)
}

func (c *RestClient) handleResponse(resp *resty.Response) (interface{}, error) {
	if resp.StatusCode() == http.StatusNoContent {
		return nil, nil
	}

	// Check for errors
	if resp.StatusCode() >= 400 {
		return nil, c.handleError(resp)
	}

	// Check content type
	contentType := resp.Header().Get("Content-Type")
	if contentType == "" || !isJSONContentType(contentType) {
		// Return as string
		return string(resp.Body()), nil
	}

	// Parse JSON
	var result interface{}
	if err := json.Unmarshal(resp.Body(), &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return result, nil
}

func (c *RestClient) handleError(resp *resty.Response) error {
	statusCode := resp.StatusCode()
	body := string(resp.Body())

	// Try to parse error message from JSON
	var errResp struct {
		Message string `json:"message"`
	}
	if err := json.Unmarshal(resp.Body(), &errResp); err == nil && errResp.Message != "" {
		body = errResp.Message
	}

	switch statusCode {
	case http.StatusUnauthorized:
		return &APIError{Code: "AUTHENTICATION_ERROR", Message: "Authentication failed: " + body}
	case http.StatusForbidden:
		return &APIError{Code: "PERMISSION_DENIED", Message: "Permission denied: " + body}
	case http.StatusNotFound:
		return &APIError{Code: "NOT_FOUND", Message: "Resource not found: " + body}
	case http.StatusBadRequest:
		return &APIError{Code: "VALIDATION_ERROR", Message: "Bad request: " + body}
	default:
		return &APIError{Code: "API_ERROR", Message: fmt.Sprintf("API error (%d): %s", statusCode, body)}
	}
}

func isJSONContentType(contentType string) bool {
	return contentType == "application/json" ||
		len(contentType) > 16 && contentType[:16] == "application/json"
}

// APIError represents an API error
type APIError struct {
	Code    string
	Message string
}

func (e *APIError) Error() string {
	return e.Message
}

// High-level API methods

// GetConfig returns the Home Assistant configuration
func (c *RestClient) GetConfig() (map[string]interface{}, error) {
	result, err := c.Get("config")
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, fmt.Errorf("unexpected response type")
}

// GetStates returns all entity states
func (c *RestClient) GetStates() ([]interface{}, error) {
	result, err := c.Get("states")
	if err != nil {
		return nil, err
	}
	if arr, ok := result.([]interface{}); ok {
		return arr, nil
	}
	return nil, fmt.Errorf("unexpected response type")
}

// GetState returns the state of a specific entity
func (c *RestClient) GetState(entityID string) (map[string]interface{}, error) {
	result, err := c.Get("states/" + entityID)
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, fmt.Errorf("unexpected response type")
}

// GetServices returns all available services
func (c *RestClient) GetServices() ([]interface{}, error) {
	result, err := c.Get("services")
	if err != nil {
		return nil, err
	}
	if arr, ok := result.([]interface{}); ok {
		return arr, nil
	}
	return nil, fmt.Errorf("unexpected response type")
}

// CallService calls a service
func (c *RestClient) CallService(domain, service string, data map[string]interface{}) (interface{}, error) {
	endpoint := fmt.Sprintf("services/%s/%s", domain, service)
	return c.Post(endpoint, data)
}

// CheckConfig validates the Home Assistant configuration
func (c *RestClient) CheckConfig() (map[string]interface{}, error) {
	result, err := c.Post("config/core/check_config", nil)
	if err != nil {
		return nil, err
	}
	if m, ok := result.(map[string]interface{}); ok {
		return m, nil
	}
	return nil, fmt.Errorf("unexpected response type")
}

// Restart restarts Home Assistant
func (c *RestClient) Restart() error {
	_, err := c.Post("services/homeassistant/restart", nil)
	return err
}

// GetErrorLog returns the error log
func (c *RestClient) GetErrorLog() (string, error) {
	result, err := c.Get("error_log")
	if err != nil {
		return "", err
	}
	if s, ok := result.(string); ok {
		return s, nil
	}
	return fmt.Sprintf("%v", result), nil
}

// GetHistory returns the state history for an entity
func (c *RestClient) GetHistory(entityID string, startTime, endTime string) ([]interface{}, error) {
	endpoint := "history/period"
	if startTime != "" {
		endpoint = "history/period/" + startTime
	}

	// Build query params
	params := ""
	if entityID != "" {
		params = "?filter_entity_id=" + entityID
	}
	if endTime != "" {
		if params == "" {
			params = "?"
		} else {
			params += "&"
		}
		params += "end_time=" + endTime
	}

	result, err := c.Get(endpoint + params)
	if err != nil {
		return nil, err
	}
	if arr, ok := result.([]interface{}); ok {
		return arr, nil
	}
	return nil, fmt.Errorf("unexpected response type")
}
